hpy/tools/autogen/debug.py

```diff
@@ -0,0 +1,164 @@
+from copy import deepcopy
+import textwrap
+from pycparser import c_ast
+from .autogenfile import AutoGenFile
+from .parse import toC, find_typedecl
+from .hpyfunc import NO_CALL
+
+class HPy_2_DHPy_Visitor(c_ast.NodeVisitor):
+    "Visitor which renames all HPy types to DHPy"
+
+    def visit_IdentifierType(self, node):
+        if node.names == ['HPy']:
+            node.names = ['DHPy']
+
+    def visit_TypeDecl(self, node):
+        if node.declname == 'ctx':
+            node.declname = 'dctx'
+        self.generic_visit(node)
+
+def funcnode_with_new_name(node, name):
+    newnode = deepcopy(node)
+    typedecl = find_typedecl(newnode)
+    typedecl.declname = name
+    return newnode
+
+def get_debug_wrapper_node(func):
+    newnode = funcnode_with_new_name(func.node, 'debug_%s' % func.ctx_name())
+    # fix all the types
+    visitor = HPy_2_DHPy_Visitor()
+    visitor.visit(newnode)
+    return newnode
+
+
+class autogen_debug_ctx_init_h(AutoGenFile):
+    PATH = 'hpy/debug/src/autogen_debug_ctx_init.h'
+
+    def generate(self):
+        lines = []
+        w = lines.append
+        # emit the declarations for all the debug_ctx_* functions
+        for func in self.api.functions:
+            w(toC(get_debug_wrapper_node(func)) + ';')
+        w('')
+        w('static inline void debug_ctx_init_fields(HPyContext dctx, HPyContext uctx)')
+        w('{')
+        for var in self.api.variables:
+            name = var.name
+            w(f'    dctx->{name} = DHPy_wrap(dctx, uctx->{name});')
+        for func in self.api.functions:
+            name = func.ctx_name()
+            w(f'    dctx->{name} = &debug_{name};')
+
+        w('}')
+        return '\n'.join(lines)
+
+
+class autogen_debug_wrappers(AutoGenFile):
+    PATH = 'hpy/debug/src/autogen_debug_wrappers.c'
+
+    NO_WRAPPER = set([
+        '_HPy_CallRealFunctionFromTrampoline',
+        'HPy_Close',
+        'HPyTuple_FromArray',
+        'HPyType_GenericNew',
+        'HPyType_FromSpec',
+        ])
+
+    def generate(self):
+        lines = []
+        w = lines.append
+        w('#include "debug_internal.h"')
+        w('')
+        for func in self.api.functions:
+            debug_wrapper = self.gen_debug_wrapper(func)
+            if debug_wrapper:
+                w(debug_wrapper)
+                w('')
+        return '\n'.join(lines)
+
+    def gen_debug_wrapper(self, func):
+        if func.name in self.NO_WRAPPER:
+            return
+        #
+        assert not func.is_varargs()
+        node = get_debug_wrapper_node(func)
+        signature = toC(node)
+        rettype = toC(node.type.type)
+        #
+        def get_params():
```

Comment:
Is there a reason `get_params` is a nested function?

Reply:
I just liked it that way. It's basically an alternative way to group a bit of logic together instead of putting a comment `compute the params`.
And putting it outside seems overkill